### PR TITLE
MainMenu Background Map scaled to Screen and minor MapGen fixes

### PR DESCRIPTION
**Pull request by SomeTroglodyte**
Thanks @alexban011 for #6983 and the inspiration to actually look it up. Extreme window aspect ratios now always get a background map that fills the window. There's a few interesting lessons in there - the virtual world size _increases_ as the window is dragged smaller... The user-chosen "resolution" has _no_ influence on the map size needed, it's all already in said world coords... The ratio in the calculation is the same as TileSet designers need to know - 64 to 56 is the same as 57.7 to 50...

'nuff rambling. Screenies over in the linked PR.

- Also fixes hardcoding what hills and mountains _are_ in applyHumidityAndTemperature - and it no longer has any business testing for hills anyway. Pure luck that didn't cause errors.
- Also fixes applyHumidityAndTemperature limitsMap containing non-base terrains. Again, luck (actually json ordering and collection order stability) that didn't lead to errors. A mod not grouping base terrains first would have triggered this.
- Also fixes MapGenerator limitsMap's not covering the allowed space - temperature -1 was excluded but allowed to be generated by the random formula, same with humility zero. D, L - who cares.
- getPerlinNoise linted because I wanted to add documentation for the return value - that it is always in the range -1..1 _inclusive_, but since I couldn't find hard evidence or supporting docs quickly I postponed that.

### CODE REVIEW

File: core/src/com/unciv/MainMenuScreen.kt [hunk 1]
```diff
@@ -75,16 +76,29 @@ class MainMenuScreen: BaseScreen() {
         ImageGetter.ruleset = RulesetCache.getVanillaRuleset()
 
         launchCrashHandling("ShowMapBackground") {
-            val newMap = MapGenerator(RulesetCache.getVanillaRuleset())
+            var scale = 1f
+            var mapWidth = stage.width / 57.735f  // TileGroupMap.groupSize * sqrt(4f/3f)
+            var mapHeight = stage.height / 50f  // TileGroupMap.groupSize
```
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/MainMenuScreen.kt:81]: Has been a hidden constant for a long time, but cross-package reuse, OK, needs more thorough cleanup.

File: core/src/com/unciv/MainMenuScreen.kt [hunk 2]
```diff
@@ -75,16 +76,29 @@ class MainMenuScreen: BaseScreen() {
...
+                mapHeight /= scale
+                scale = min(scale, 20f)
+            }
+
+            val mapRuleset = RulesetCache.getVanillaRuleset()
+            val newMap = MapGenerator(mapRuleset)
                     .generateMap(MapParameters().apply {
                         shape = MapShape.rectangular
-                        mapSize = MapSizeNew(MapSize.Small)
+                        mapSize = MapSizeNew(mapWidth.toInt() + 1,mapHeight.toInt() + 1)
```
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/MainMenuScreen.kt:93]: And I'm the kotlinlang.org#horizontal-space pointing "grey man" around here :see_no_evil:

File: core/src/com/unciv/logic/map/mapgenerator/MapGenerator.kt [hunk 3]
```diff
@@ -45,13 +46,26 @@ class MapGenerator(val ruleset: Ruleset) {
...
         constructor(terrain: Terrain, unique: Unique)
                 : this(terrain,
-            unique.params[0].toFloat(), unique.params[1].toFloat(),
-            unique.params[2].toFloat(), unique.params[3].toFloat())
-        /** checks if both [temperature] and [humidity] satisfy their ranges (>From, <=To) */
+            unique.params[0].toFloatMakeInclusive(-1f), unique.params[1].toFloat(),
+            unique.params[2].toFloatMakeInclusive(0f), unique.params[3].toFloat())
+
+        /** Checks if both [temperature] and [humidity] satisfy their ranges (>From, <=To)
+         *  Note the lowest allowed limit has been made inclusive (temp -1 nudged down by 1 [Float.ulp], humidity at 0)
```
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/logic/map/mapgenerator/MapGenerator.kt:53]: Problem is, the ranges are written in json as from A to B ignoring any open/closed range distinction. The input values are on a [-1,1] closed range. json ranges are allowed to be what they want inside that (as long as any point in the random input space is covered) and we want, say, a json definition of [-1..0] -> Grass; [0..1] -> Desert to be deterministic, repeatable and commutative at the zero. So json ranges need to be open one end closed the other. That leaves _ONE_ possible input value out, unless every modder (and we into vanilla) writes the ulp into their json. We _had_ Grass near the poles in extreme settings. I decided to leave the matches function as is - open/closed - but to auto-tweak the json ranges so every lowest end gets nudged down by one ulp. The alternative - make matches do closed/closed comparison - would be an imperfection probably nobody would notice but me. That nudge isn't expensive, so why not?

> assumption towards some kind of values

Yes: The Perlin generators deliver -1..1 inclusive.

File: core/src/com/unciv/logic/map/mapgenerator/MapGenerator.kt [hunk 4]
```diff
@@ -463,11 +477,15 @@ class MapGenerator(val ruleset: Ruleset) {
 
         // List is OK here as it's only sequentially scanned
         val limitsMap: List<TerrainOccursRange> =
-            ruleset.terrains.values.flatMap {
-                it.getGenerationConditions()
-            }
+            ruleset.terrains.values.filter { it.type == TerrainType.Land }
```
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/logic/map/mapgenerator/MapGenerator.kt:480]: This is in the step deciding on snow/tundra/grass/plain/desert for land. Coast/Ocean is done elsewhere. Yes, modding more Water base terrains doesn't make much sense, still much hardcoding. Besides, Land is _base_ terrain, Ice a terrain _feature_.

**Comment by SomeTroglodyte** [comment]
Gimme a minute or a hundred of them. I'm stunned by [this beauty](https://github.com/libgdx/libgdx/blob/84fdb1c1e58ea79931139f355dab7aaedc2abc75/backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/DefaultLwjgl3Input.java#L363) and right now unable to code.

File: core/src/com/unciv/MainMenuScreen.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/MainMenuScreen.kt]: OK, that cleanup is too advanced for me. So many places that _seem_ to relate but then do not match exactly. Perhaps that's why borders are displaced when one looks closely... But TileGroupMap.groupSize as public const looks good.

[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/MainMenuScreen.kt]: P.S. yes turns out 57.7 was the wrong pen & paper result I copied - I had both the advance (43.3) and the diagonal scribbled then looked a little to diagon alley...

File: core/src/com/unciv/MainMenuScreen.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexban011) [reply core/src/com/unciv/MainMenuScreen.kt]: yeah thats how I forget about the spaces. Would be cool if studio had a shortcut to toggle those on and off. Would make cleanup before pushing easier.

[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/MainMenuScreen.kt]: It has.. It's just not too aggressive
<details><summary>Settings</summary>

image
image
</details>

[the remark this answers is not in the document]
Comment (alexban011) [reply core/src/com/unciv/MainMenuScreen.kt]: Sorry, meant to say toggle shortcut for the name parameter derations

File: core/src/com/unciv/logic/map/mapgenerator/MapGenerator.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/logic/map/mapgenerator/MapGenerator.kt:53]: Current: ( Input [-1..1] Definitions ]-1..1] ). PR: ( Input [-1..1] Definitions [-1..1] ). Your suggestion: ( Input [-1..1] Definitions [-1..1] with ambiguity). Now roll a [d65536](https://xkcd.com/2626) to see which we keep.

**Comment by yairm210** [comment]
Conflicts 🙃

**Comment by SomeTroglodyte** [comment]
> :upside_down_face: 

Of course, the pan accelerator PR, unavoidable. Studio tied up ATM.

**Pull request merged**